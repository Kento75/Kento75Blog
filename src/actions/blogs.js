import database from '../firebase/firebase';

// ADD_BLOG
export const ADD_BLOG = 'ADD_BLOG';

export const addBlog = (blog) => ({
	type: ADD_BLOG,
	blog
});

export const startAddBlog = (blogData = {}) => {
	return (dispatch, getState) => {
		const uid = getState().auth.uid;
		const {
			title = '',
			body = '',
			createdAt = 0
		} = blogData;

		const blog = { title, body, createdAt };

		return database.ref(`users/${uid}/blogs`).push(blog).then((ref) => {
			dispatch(addBlog({ id: ref.key, ...blog }));
		});
	};
};

// REMOVE_BLOG
export const REMOVE_BLOG = 'REMOVE_BLOG';

export const removeBlog = ({ id } = {}) => ({
	type: REMOVE_BLOG,
	id
});

export const startRemoveBlog = ({ id }) => {
	return (dispatch, getState) => {
		const uid = getState().auth.uid;

		return database.ref(`users/${uid}/blogs/${id}`).remove().then(() => {
			dispatch(removeBlog({ id }));
		});
	};
};

// EDIT_BLOG
export const EDIT_BLOG = 'EDIT_BLOG';

export const editBlog = (id, updates) => ({
	type: EDIT_BLOG,
	id,
	updates
});

export const startEditBlog = (id, updates) => {
	return (dispatch, getState) => {
		const uid = getState().auth.uid;

		return database.ref(`users/${uid}/blogs/${id}`).update(updates).then(() => {
			dispatch(editBlog(id, updates));
		});
	};
};

// SET_BLOG
export const SET_BLOG = 'SET_BLOG';

export const setBlogs = (blogs) => ({
	type: SET_BLOG,
	blogs
});

export const startSetBlogs = () => {
	return (dispatch, getState) => {
		const uid = getState().auth.uid;

		return database.ref(`users/${uid}/blogs`).once('value').then((snapshot) => {
			const blogs = [];
			snapshot.forEach((blog) => {
				blogs.push({
					id: blog.key,
					...blog.val()
				});
			});
			dispatch(setBlogs(blogs));
		});
	};
};

// GET_BLOG - To fetch the blog based on id
export const startGetBlog = ({ id }) => {
	return () => {
		return database.ref(`users`).once('value').then((snapshot) => {
			let blog;
			snapshot.forEach((child) => {
				blog = child.val().blogs[id];
			});
			return blog;
		});
	};
};