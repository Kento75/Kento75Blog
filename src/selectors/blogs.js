export default (blogs, { text, sortBy }) => {
	return blogs.filter((blog) => {
		const titleMatch = blog.title.toLowerCase().includes(text.toLowerCase());

		return titleMatch;
	}).sort((a, b) => {
		if (sortBy === 'date') {
			return a.createdAt < b.createdAt ? 1 : -1;
		}
	});
};