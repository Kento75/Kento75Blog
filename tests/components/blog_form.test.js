import React from 'react';
import { shallow } from 'enzyme';
import moment from 'moment';

import BlogForm from '../../src/components/blog_form';
import blogs from '../fixtures/blogs';

test('should render BlogForm correctly', () => {
	const wrapper = shallow(<BlogForm />);
	expect(wrapper).toMatchSnapshot();
});

test('should render BlogForm correctly with blog data', () => {
	const wrapper = shallow(<BlogForm blog={blogs[1]} />);

	expect(wrapper).toMatchSnapshot();
});

test('should render error for invalid form submission', () => {
	const wrapper = shallow(<BlogForm />);
	wrapper.find('form').simulate('submit', {
		preventDefault: () => {}
	});

	expect(wrapper.state('error').length).toBeGreaterThan(0);
});

test('should set title on input change', () => {
	const wrapper = shallow(<BlogForm />);
	const value = 'blah';

	wrapper.find('input').at(0).simulate('change', {
		target: { value }
	});

	expect(wrapper.state('title')).toBe(value);
});

test('should set body on textarea change', () => {
	const wrapper = shallow(<BlogForm />);
	const value = '<p>blah</p>';
	wrapper.find('Quill').at(0).simulate('change', value);

	expect(wrapper.state('body')).toBe(value);
});

test('should set title on input change', () => {
	const wrapper = shallow(<BlogForm />);
	const value = 'blah';

	wrapper.find('input').at(0).simulate('change', {
		target: { value }
	});

	expect(wrapper.state('title')).toBe(value);
});

test('should call onSubmit prop for valid form submission', () => {
	const onSubmitSpy = jest.fn(); //Test spy
	const wrapper = shallow(<BlogForm blog={blogs[1]} onSubmit={onSubmitSpy} />);

	wrapper.find('form').simulate('submit', {
		preventDefault: () => {}
	});
	expect(wrapper.state('error')).toBe('');
	expect(onSubmitSpy).toHaveBeenCalledWith({
		title: blogs[1].title,
		body: blogs[1].body,
		createdAt: blogs[1].createdAt,
	});
});