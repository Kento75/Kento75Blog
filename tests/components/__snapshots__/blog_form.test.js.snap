// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render BlogForm correctly 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="Post Title"
    type="text"
    value=""
  />
  <Quill
    modules={Object {}}
    onChange={[Function]}
    theme="snow"
    value=""
  />
  <input
    className="button"
    type="submit"
    value="Save Post"
  />
</form>
`;

exports[`should render BlogForm correctly with blog data 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="Post Title"
    type="text"
    value="Write a Angular app"
  />
  <Quill
    modules={Object {}}
    onChange={[Function]}
    theme="snow"
    value="Directives, Lazy Load"
  />
  <input
    className="button"
    type="submit"
    value="Save Post"
  />
</form>
`;
